### PR TITLE
Add place order confirmation page to express checkout

### PR DESCRIPTION
**Pull request by sergeysikachev**
- Developer(s): @sergeysikachev

- Peer Reviewer:

- Ticket: -

- Project PR: -

- Academy PR: - (for feature documentation and if needed migration guides and project integration guide)

#### Please confirm

- [x] No new OS components - or they have been approved by legal department
- [x] All new or heavily changed classes get an "A" rating (Scrutinizer), except Facades, Factories, QueryContainers and Tests

#### Documentation

- [x] Functional documentation provided or in progress?
- [x] Integration guide for projects provided or not needed?
- [x] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraints         |
   :--------------------- | :------------------------ | :--------------------- |
   Payone                | minor                 |                       |

#### Release Notes

Features:

* ExpressCheckout logic has been made StepEngine based.
* User is made to be redirected to summary page from payone instead of instant order placing.
* Summary page has shipment method select, so user can specify it after coming back from PayPal.

-----------------------------------------

#### Module Payone

_**Minor:** Backwards-compatible bug fix_

Def of done (by responsible developer):
- [x] All changes are backward-compatible. Outdated code is marked as deprecated
- [x] The change is isolated and not mixed with any cleanups or other changes
- [x] New code fits to the architecture rules
- [x] All new or changed business logic is covered by functional tests (in Zed business layer)

##### Change log

Features:

* ExpressCheckout logic has been made StepEngine based.
* User is made to be redirected to summary page from payone instead of instant order placing.
* Summary page has shipment method select, so user can specify it after coming back from PayPal.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
     "spryker/refund": "^4.0.0 || ^5.0.0",
     "spryker/sales": "^5.0.0 || ^6.0.0 || ^7.0.0",
     "spryker/symfony": "^3.0.0",
-    "spryker/product-bundle": "^3.0.0"
+    "spryker/product-bundle": "^3.0.0",
```
Comment (dereuromark) [inline composer.json]: Tip: If you add them via "composer require" command, they will stay sorted.

File: src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+class StepFactory extends AbstractFactory
+{
+
+    /**
+     * @param \Spryker\Yves\StepEngine\Process\StepCollectionInterface $stepCollection
+     *
+     * @return \Spryker\Yves\StepEngine\Process\StepEngine
+     */
+    public function createStepEngine(StepCollectionInterface $stepCollection)
```
Comment (profuel) [inline src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php]: why you need this argument?
I would use directly method `createStepCollection`.
It still allows you to modify functionality on project level

File: src/SprykerEco/Yves/Payone/Plugin/Provider/PayoneControllerProvider.php [hunk 3]
```diff
@@ -18,8 +18,11 @@ class PayoneControllerProvider extends YvesControllerProvider
     const EXPRESS_CHECKOUT_BUTTON = 'payone-checkout-with-paypal-button';
     const EXPRESS_CHECKOUT_INIT = 'payone-paypal-express-checkout-start';
     const EXPRESS_CHECKOUT_SUCCESS = 'payone-paypal-express-checkout-success';
+    const EXPRESS_CHECKOUT_SUMMARY = 'payone-paypal-express-checkout-summary';
```
Comment (profuel) [inline src/SprykerEco/Yves/Payone/Plugin/Provider/PayoneControllerProvider.php]: Maybe worth renaming all routes with prefix `ROUTE_` ?

File: src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    /**
+     * @return \Spryker\Client\Shipment\ShipmentClientInterface
+     */
+    public function getShipmentClient()
+    {
+        return $this->getProvidedDependency(PayoneDependencyProvider::CLIENT_SHIPMENT);
+    }
+
+    /**
+     * @return \SprykerEco\Yves\Payone\Handler\ExpressCheckout\QuoteHydrator
```
Comment (profuel) [inline src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php]: interface is missing. Please run code & arch sniffer

File: src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    /**
+     * @return \Symfony\Component\Routing\Generator\UrlGeneratorInterface
+     */
+    protected function getUrlGenerator()
+    {
+        return $this->getApplication()['url_generator'];
+    }
+
+    /**
+     * @return mixed
```
Comment (profuel) [inline src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php]: not allowed, please use correct type-hint

File: src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    {
+        return $this->getProvidedDependency(PayoneDependencyProvider::PLUGIN_APPLICATION);
+    }
+
+    /**
+     * @return \Spryker\Yves\ProductBundle\Grouper\ProductBundleGrouper
+     */
+    protected function createProductBundleGrouper()
+    {
+        return new ProductBundleGrouper();
```
Comment (profuel) [inline src/SprykerEco/Yves/Payone/Checkout/Process/StepFactory.php]: Please avoid this.
It couples code with product bundles.
For AmazonPay we've moved code to project level

File: src/SprykerEco/Yves/Payone/Checkout/Process/Steps/InitExpressCheckoutStep.php [hunk 7]
```diff
@@ -33,61 +35,106 @@ class ExpressCheckoutHandler implements ExpressCheckoutHandlerInterface
...
-     * @return \Symfony\Component\HttpFoundation\RedirectResponse
+     * Precondition here is required to exit the checkout and go to cart.
+     *
+     * @param \Spryker\Shared\Kernel\Transfer\AbstractTransfer $dataTransfer
+     *
+     * @return bool
+     */
+    public function preCondition(AbstractTransfer $dataTransfer)
+    {
+        $payment = $dataTransfer->getPayment();
```
Comment (profuel) [inline src/SprykerEco/Yves/Payone/Checkout/Process/Steps/InitExpressCheckoutStep.php]: Please add correct type-hint for `$dataTransfer` in all functions and steps

**Comment by sergeysikachev** [comment]
Pull request is abandoned because It was decided to avoid step engine usage in module.

**Pull request closed without merging**